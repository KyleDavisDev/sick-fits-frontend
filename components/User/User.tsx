import * as React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { ReactElement } from "react";

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      cart {
        created
        id
        items {
          id
          quantity
          item {
            id
            price
            image
            title
            description
          }
        }
      }
    }
  }
`;

export interface IUserProps {
  children(data: any): ReactElement;
}

const User: React.FunctionComponent<IUserProps> = props => {
  return (
    <Query {...props} query={CURRENT_USER_QUERY}>
      {payload => {
        return props.children(payload);
      }}
    </Query>
  );
};

export default User;
