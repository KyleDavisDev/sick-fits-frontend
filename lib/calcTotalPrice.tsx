export default function(items) {
  return items.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
