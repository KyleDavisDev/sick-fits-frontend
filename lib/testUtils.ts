import casual from "casual";
import { ReactWrapper } from "enzyme";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  created: casual.unix_time,
  id: casual.uuid,
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = (overrides?: any) => ({
  __typename: "CartItem",
  id: casual.uuid,
  created: casual.unix_time,
  quantity: casual.integer(0, 15),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

const fakeCart = (overrides?: any) => ({
  __typename: "Cart",
  id: casual.uuid,
  items: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
  user: fakeUser(),
  created: casual.unix_time,
  updated: casual.unix_time,
  isActive: true,
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  store: any;
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeCart,
  fakeOrder,
  fakeOrderItem,
  simulateInputChange
};
