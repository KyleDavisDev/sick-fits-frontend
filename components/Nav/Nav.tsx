import * as React from "react";
import Link from "next/link";
import { NavStyles } from "./NavStyles";
import User from "../User/User";

interface INavProps {}

const Nav: React.FunctionComponent<INavProps> = props => {
  return (
    <User>
      {({ data: { me } }) => {
        return (
          <NavStyles>
            <Link href="/Shop">
              <a>Shop</a>
            </Link>

            {me && (
              <>
                <Link href="/Sell">
                  <a>Sell</a>
                </Link>
                <Link href="/Orders">
                  <a>Orders</a>
                </Link>
                <Link href="/Me">
                  <a>Account</a>
                </Link>
              </>
            )}
            {!me && (
              <Link href="/Signup">
                <a>Sign In</a>
              </Link>
            )}
          </NavStyles>
        );
      }}
    </User>
  );
};

export default Nav;
