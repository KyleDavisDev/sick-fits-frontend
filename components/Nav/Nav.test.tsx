import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { CURRENT_USER_QUERY } from "../User/User";
import Nav from "./Nav";
import { fakeUser } from "../../lib/testUtils";

describe("<Nav />", () => {
  const notSignedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { me: null } }
    }
  ];

  const signedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { me: fakeUser() } }
    }
  ];

  it("renders", () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav></Nav>
      </MockedProvider>
    );

    expect(wrapper).toBeTruthy();
  });

  it("matches snapshot", async () => {
    // check signed out user first
    let wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav></Nav>
      </MockedProvider>
    );

    // allow component to load/update
    await wait();
    wrapper.update();

    let nav = wrapper.find('[data-test="nav"]');

    // check snapshot
    expect(nav).toMatchSnapshot();

    // mount component w/ signed in user
    wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav></Nav>
      </MockedProvider>
    );

    // allow component to load/update
    await wait();
    wrapper.update();

    nav = wrapper.find('[data-test="nav"]');

    // check snapshot
    expect(nav).toMatchSnapshot();
  });

  it("renders signed out nav when user is not logged in", async () => {
    // mount component w/ signed in user
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav></Nav>
      </MockedProvider>
    );

    // allow component to load/update
    await wait();
    wrapper.update();

    const nav = wrapper.find('[data-test="nav"]');

    // check both buttons
    expect(nav.find("Link").length).toEqual(2);
    // check first button text
    expect(
      nav
        .find("Link")
        .at(0)
        .text()
    ).toEqual("Shop");
    // check second button text
    expect(
      nav
        .find("Link")
        .at(1)
        .text()
    ).toEqual("Sign In");
  });
});
