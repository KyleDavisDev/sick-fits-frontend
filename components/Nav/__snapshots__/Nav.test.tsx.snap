// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> matches snapshot 1`] = `
<ul
  className="sc-bdVaJa VvFAz"
  data-test="nav"
>
  <Link
    href="/shop"
  >
    <a
      href="/shop"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/Signup"
  >
    <a
      href="/Signup"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Sign In
    </a>
  </Link>
</ul>
`;

exports[`<Nav /> matches snapshot 2`] = `
<ul
  className="sc-bdVaJa VvFAz"
  data-test="nav"
>
  <Link
    href="/shop"
  >
    <a
      href="/shop"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/sell"
  >
    <a
      href="/sell"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Sell
    </a>
  </Link>
  <Link
    href="/orders"
  >
    <a
      href="/orders"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Orders
    </a>
  </Link>
  <Link
    href="/me"
  >
    <a
      href="/me"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Account
    </a>
  </Link>
  <SignOut>
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": Object {
                "kind": "Name",
                "value": "SIGNOUT_MUTATION",
              },
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "signOut",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "message",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 69,
            "start": 0,
          },
        }
      }
      refetchQueries={
        Array [
          Object {
            "query": Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": Object {
                    "kind": "Name",
                    "value": "CURRENT_USER_QUERY",
                  },
                  "operation": "query",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "me",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "email",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "permissions",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "cart",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "created",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "items",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "id",
                                          },
                                          "selectionSet": undefined,
                                        },
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "quantity",
                                          },
                                          "selectionSet": undefined,
                                        },
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "item",
                                          },
                                          "selectionSet": Object {
                                            "kind": "SelectionSet",
                                            "selections": Array [
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "id",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "price",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "image",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "title",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "description",
                                                },
                                                "selectionSet": undefined,
                                              },
                                            ],
                                          },
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 327,
                "start": 0,
              },
            },
          },
        ]
      }
    >
      <button
        onClick={[Function]}
      >
        Sign Out
      </button>
    </Mutation>
  </SignOut>
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": undefined,
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [
                    Object {
                      "arguments": Array [],
                      "kind": "Directive",
                      "name": Object {
                        "kind": "Name",
                        "value": "client",
                      },
                    },
                  ],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "toggleCart",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
            "variableDefinitions": Array [],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 41,
          "start": 0,
        },
      }
    }
  >
    <button
      onClick={[Function]}
    >
      My Cart
      <CartCount
        count={0}
      >
        <styled.span>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htoDjs",
                  "isStatic": false,
                  "lastClassName": "kfwyaH",
                  "rules": Array [
                    "
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }

  /* initial state of the entered dot */
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }

  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htoDjs",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <span
              className="sc-htoDjs kfwyaH"
            >
              <TransitionGroup
                childFactory={[Function]}
                component="div"
              >
                <div>
                  <CSSTransition
                    className="count"
                    classNames="count"
                    in={true}
                    key=".$0"
                    onExited={[Function]}
                    timeout={
                      Object {
                        "enter": 400,
                        "exit": 400,
                      }
                    }
                    unmountOnExit={true}
                  >
                    <Transition
                      appear={false}
                      className="count"
                      enter={true}
                      exit={true}
                      in={true}
                      mountOnEnter={false}
                      onEnter={[Function]}
                      onEntered={[Function]}
                      onEntering={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onExiting={[Function]}
                      timeout={
                        Object {
                          "enter": 400,
                          "exit": 400,
                        }
                      }
                      unmountOnExit={true}
                    >
                      <styled.div
                        className="count"
                      >
                        <StyledComponent
                          className="count"
                          forwardedComponent={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": Array [],
                              "componentStyle": ComponentStyle {
                                "componentId": "sc-gzVnrw",
                                "isStatic": false,
                                "lastClassName": "edyIo",
                                "rules": Array [
                                  "
  background-color: ",
                                  [Function],
                                  ";
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: \\"tnum\\";
  font-variant-numeric: tabular-nums;
",
                                ],
                              },
                              "displayName": "styled.div",
                              "foldedComponentIds": Array [],
                              "render": [Function],
                              "styledComponentId": "sc-gzVnrw",
                              "target": "div",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                        >
                          <div
                            className="sc-gzVnrw edyIo count"
                          >
                            0
                          </div>
                        </StyledComponent>
                      </styled.div>
                    </Transition>
                  </CSSTransition>
                </div>
              </TransitionGroup>
            </span>
          </StyledComponent>
        </styled.span>
      </CartCount>
    </button>
  </Mutation>
</ul>
`;
